### ISSUE TITLE
Tracking Issue for generic `NonZero`

### ISSUE THREAD

**Issue by reitermarkus** [rust-lang/rust#120257]
Feature gate: `#![feature(generic_nonzero)]`

This is a tracking issue for replacing the distinct `NonZero*` types with a generic `NonZero<T>` type. This allows using `NonZero` with FFI type aliases instead of having weird type names like `NonZero_c_ulonglong`.

This replaces the following tracking issues:

- https://github.com/rust-lang/rust/issues/95157
- https://github.com/rust-lang/rust/issues/82363

### Public API

```rust
use core::num::{NonZero, NonZeroU8};

assert_eq!(NonZero::new(33u8), NonZeroU8::new(33));
```

### Steps / History

- [x] Initial implementation attempt: https://github.com/rust-lang/rust/pull/100428
- [x] https://github.com/rust-lang/rust/pull/118665
- [x] https://github.com/rust-lang/rust/pull/119990
- [x] https://github.com/rust-lang/rust/pull/120160
- [x] https://github.com/rust-lang/rust/pull/120165
- [x] https://github.com/rust-lang/rust/pull/120244
- [x] https://github.com/rust-lang/rust/pull/120521
- [x] https://github.com/rust-lang/rust/pull/120809
- [x] https://github.com/rust-lang/rust/pull/120563
- [x] https://github.com/rust-lang/rust/pull/120486
- [ ] Make other methods generic. (optional, can happen after stabilization)
  - [x] https://github.com/rust-lang/rust/pull/121241
- [x] https://github.com/rust-lang/rust/pull/121277
- [x] https://github.com/rust-lang/rust/pull/121454
- [x] https://github.com/rust-lang/rust/pull/121461
- [x] https://github.com/rust-lang/rust/pull/121850
- [x] Final comment period (FCP)[^1]
  - [x] https://github.com/rust-lang/rust/pull/121885
- [ ] Stabilization
  - [ ] https://github.com/rust-lang/rust/pull/124230
  - [ ] Update examples to use `NonZero`.

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by reitermarkus** [comment]
@dtolnay, given the documentation issue, making `NonZero::new` and other constructors generic can really only be done during/after stabilization, so the only blocker I can see right now is stabilization itself. Can this enter FCP, or do we want to wait for https://github.com/rust-lang/rust/pull/120486?

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

Providing NonZeroT types as 12 separate structs ([`NonZeroI32`](https://doc.rust-lang.org/1.76.0/std/num/struct.NonZeroI32.html) and friends) as opposed to a single generic `NonZero<T>` type is _"[the gravest API mistake Rust has ever made](https://github.com/rust-lang/rust/issues/82363#issuecomment-921513910)"_, by some accounts.

In https://github.com/rust-lang/rust/issues/82363 we agreed to pursue generic `NonZero<T>`, and in doing so, it would straightforwardly become usable with C integers as `NonZero<core::ffi::c_int>` rather than us needing to stabilize a family of awkwardly named `NonZero_c_int` type aliases.

That work has come to fruition and I propose that we are ready to stabilize `NonZero<T>`.

As of Rust 1.77, the NonZeroT types have been redefined like this:

```rust
pub type NonZeroI32 = NonZero<T>;
...
```

where `NonZeroI32` is already stable but `NonZero` is not; it is only usable through the set of 12 stable type aliases. The API of the type aliases is identical to the API of the original 12 NonZeroT structs. All code written against NonZeroT when they were structs continues to work with NonZeroT being aliases to NonZero\<T\>.

The **new** surface area being proposed for stabilization is:

1. The ability to spell `core::num::NonZeroI32` as `core::num::NonZero<i32>`.
2. A straightforward way to write `NonZero<c_int>`. Previously, expressing this in stable Rust was possible only by defining your own type aliases as follows, and using `<c_int as MyZeroablePrimitive>::NonZero`:
    ```rust
    trait MyZeroablePrimitive { type NonZero; }
    impl MyZeroablePrimitive for i16 { type NonZero = NonZeroI16; }
    impl MyZeroablePrimitive for i32 { type NonZero = NonZeroI32; }
    impl MyZeroablePrimitive for i64 { type NonZero = NonZeroI64; }
    ```
3. A subset of the types' associated functions (currently limited to `new` and `new_unchecked` and `get`) becoming associated with NonZero, for example as `NonZero::new(1i32)` for what previously had to be written as `NonZeroI32::new(1)`. The latter will continue to work.
4. A subset of the types' trait functions becoming associated with NonZero: this applies to trait functions from the `Debug`, `Display`, `Binary`, `Octal`, `LowerHex`, `UpperHex`, `Clone`, `PartialEq`, `PartialOrd`, `Ord`, `Hash`, `From`, `BitOr`, and `BitOrAssign` traits. For example one could in principle write `NonZero::fmt(&nonzero_i32, f)` if they wanted, instead of `NonZeroI32::fmt(&nonzero_i32, f)` or `nonzero_i32.fmt(f)`.

Other surface area not being proposed for stabilization right now is:

5. The ability to write code that refers to `NonZero<T>` in which `T` is a generic type parameter, as opposed to a concrete type, with any less effort than it required before. For example, being able to write the following is not included in this FCP.
    ```rust
    fn generic<T>(nonzero: NonZero<T>) -> T
    where
        T: From<NonZero<T>>,
    {
        T::from(nonzero)
    }
    ```
    You would instead perhaps write:
    ```rust
    trait NonZeroPrimitive: Sized { type Primitive: From<Self>; }
    impl NonZeroPrimitive for NonZero<i32> { type Primitive = i32; }

    fn generic<NZ: NonZeroPrimitive>(nonzero: NZ) -> NZ::Primitive {
        NZ::Primitive::from(nonzero)
    }
    ```
    which of course you can already write on stable today without `NonZero`.
6. The rest of items 3 and 4, i.e. other associated functions which remain associated with NonZeroT only. For example you cannot write `NonZero::ilog2(nonzero_i32)`, only `NonZeroI32::ilog2(nonzero)` as before.

Limitation 5 pertains to stabilization of the sealed [`ZeroablePrimitive`](https://doc.rust-lang.org/beta/std/num/trait.ZeroablePrimitive.html) trait for use as a trait bound, which is under a different feature than `NonZero` and is not proposed for stabilization in this FCP. We may or may not want to pursue making it stable in the future, although it may need to look substantially different than its current form for that.

Limitation 6 exists to ensure `NonZero<T>`'s API is not reliant on any `T: const Trait` bounds. The implementation looks like this:

```rust
#[repr(transparent)]
#[rustc_layout_scalar_valid_range_start(1)]
#[rustc_nonnull_optimization_guaranteed]
pub struct NonZero<T: ZeroablePrimitive>(T);

// some functions which are generically defined, and associated with NonZero
impl<T: ZeroablePrimitive> NonZero<T> {
    pub const fn new(n: T) -> Option<Self> {...}
    ...
}

// some functions which are separately defined for each concrete instantiation
impl NonZero<i32> {
    pub const fn ilog2(self) -> u32 {...}
    ...
}
```

Currently we are holding off on anything that would commit to making `ZeroablePrimitive` not just a marker trait, such as:

```rust
impl<T> NonZero<T> {
    pub const fn ilog2(self) -> u32
    where
        T: const ZeroablePrimitive, // ???
    {
        <T as ZeroablePrimitive>::do_ilog2(self) // ???
    }
}
```

### Timeline

I am interested in recommending a more accelerated stabilization for `NonZero<T>` than the average "leisurely" library feature.

As of Rust 1.77, **stable** NonZeroT are type aliases for **unstable** NonZero\<T\> which is an unusual and disorienting state. The standard library's rustdoc when 1.77 is released will be rendering `NonZero<T>` in various stable APIs where that type cannot be named by a user of stable Rust. I recommend that we attempt to avoid leaving things in this state for too long.

`NonZero<T>` is being used in rustc and miri since https://github.com/rust-lang/rust/pull/120486, and works as intended. Any breakage from the refactor (NonZeroT going from struct to type alias) would almost certainly have surfaced over the last 3 weeks it's been in nightly. It is in beta now and will be covered by the 1.77-beta crater runs. We can act now to approve a stabilization for 1.78 or 1.79, and in the almost inconceivable case that we become aware of a deal-breaker upon the release of 1.77-stable, there would still be time to bail out of the stabilization then.

### Acknowledgements

- @reitermarkus for driving nearly all of the implementation work for NonZero\<T\> over a period of 1.5 years.
- @Nilstrieb for making it one of their first actions upon being granted library reviewer permission to unstick the work on NonZero\<T\>, which had been stalled on code review.
- @kennytm for first suggesting that NonZero\<T\> be implemented as a struct, and rejecting a less good suggestion by yours truly. (https://github.com/rust-lang/rust/issues/82363#issuecomment-906658138)
- @AngelicosPhosphoros for writing codegen tests for NonZeroT, which we made sure do not regress with NonZero\<T\>.
- @scottmcm for pinpointing compiler performance impact from NonZero\<T\>, guidance solving it, and verifying the fix (making generic `new` generate less MIR).

**Comment by ebkalderon** [comment]
This may be considered off-topic (please feel free to mark it as such), but I really like the direction the RFC and this tracking issue has taken so far. Has there been any thought to possibly starting a similar effort to (re)introduce a similar `Atomic<T>` type? I see some parallels between this effort to introduce a `NonZero<T>` type into `core` and a very similar closed RFC from 2016 to introduce an `Atomic<T>` type (https://github.com/rust-lang/rfcs/pull/1505). If this RFC gets accepted, would there be any appetite to attempt a similar effort for an `Atomic<T>` type as well, at a later date?

**Comment by reitermarkus** [comment]
> Has there been any thought to possibly starting a similar effort to (re)introduce a similar `Atomic<T>` type?

I have thought about it, and it definitely makes sense to me. It may be tricker to implement given there is `AtomicBool` and `AtomicPtr` and all are `repr(C)` instead of `repr(transparent)`.

**Comment by HadrienG2** [comment]
Another thing which makes the atomic type situation trickier is that not all atomic types have the same API (e.g. AtomicUsize has fetch_add, AtomicBool doesn't).

**Comment by kennytm** [comment]
The "different API" scenario is exactly the same as NonZero (e.g. only NonZeroIxx has the `is_negative()` function and only NonZeroUxx has the `ilog2()` function). 

There could be just one impl block per type e.g.

```rust
impl Atomic<u64> {
    #[cfg(target_has_atomic = "64")]
    fn fetch_add(&self, val: u64, ordering: Ordering) -> u64 { ... }
}
```

because of the `#[cfg]` I don't think this can be done in a more generic way (`impl<T: Integer> Atomic<T>`).

**Comment by dtolnay** [comment]
Independent of whether `ZeroablePrimitive` is unstable, it is correctly sealed in a way that cannot be circumvented by enabling nightly features, so there cannot be impls other than the ones provided by libcore.

**Comment by Skgland** [comment]
If the trait encodes a correctness/soundness requirement, shouldn't it be an `unsafe trait`?

**Comment by reitermarkus** [comment]
@Skgland, yes, I actually wanted to make it an `unsafe trait` but then forgot about it.

**Comment by RalfJung** [comment]
So we are relying on layout computation never firing even when the user writes types like `NonZero<[u8; 3]>`, based on the missing trait implementations for that type.

@oli-obk can you think of any ways that could cause a problem? Are there cases where we keep going with a type like `NonZero<[u8; 3]>` despite its missing trait impls, and then compute a layout somewhere? That would be an ICE reachable on stable if this gets accepted.

**Comment by oli-obk** [comment]
I don't think we should land this at this time.

The attribute is not part of the type system. It is currently being removed in favor of unstable pattern types. These are a more principled approach, but currently incompatible with these generic types (as you cannot specify a range pattern for generic types in matches either )

**Comment by Nilstrieb** [comment]
I think this concern can be resolved by making `NonZero` a `repr(transparent)` wrapper over an associated type on `NonZeroablePrimitive` which uses the repr attributes (or pattern types). This needs to be done before stabilization of course.
@rfcbot concern generic-layout-computation

**Comment by m-ou-se** [comment]
Registering the concern from the comment above:

@rfcbot concern generic-layout-computation

**Comment by reitermarkus** [comment]
> That's also nice in that if it's `struct NonZero<T>(T::AssocType);` it'll avoid the "why are you bounding in the type definition?"

Do you not still need to have a bound to know which trait the `AssocType` comes from?

**Comment by reitermarkus** [comment]
> I think this concern can be resolved by making `NonZero` a `repr(transparent)` wrapper over an associated type on `NonZeroablePrimitive` which uses the repr attributes (or pattern types). This needs to be done before stabilization of course.

I [tried this now](https://github.com/rust-lang/rust/pull/121885), but for some reason `Option<NonZero<_>>` ends up as “not FFI-safe”. Not sure if this is just a diagnostics bug.

**Comment by reitermarkus** [comment]
The concern is resolved now that https://github.com/rust-lang/rust/pull/121885 is merged.

**Comment by reitermarkus** [comment]
Can someone remove the concern here, please?

**Comment by m-ou-se** [comment]
@rfcbot resolve generic-layout-computation

**Comment by tmccombs** [comment]
Just a thought:

I think this could potentially be made user-implementable if there was an API that looked like:

```rust
pub trait Zeroable: Copy + StructuralPartialEq + Sized {
  const ZERO: Self;
}

pub struct NonZero<T: Zeroable>(T);

impl<T: Zeroable> NonZero<T> {
  const fn new(v: T) -> Option<Self> {
    if (v == T::ZERO) {
      None
    } else {
      Some(NonZero(v))
    }
  }
  // ...
}
```

**Comment by tmccombs** [comment]
Oh, I forgot to mention that the compiler would need to be able to treat rhe value of ZERO as a niche for the NonZero type. I'm not sure how feasible that is.

**Comment by Alonely0** [comment]
> Just a thought:
> 
> I think this could potentially be made user-extensable if there was an API that looked like:
> 
> ```rust
> pub trait Zeroable: Copy + StructuralPartialEq + Sized {
>   const ZERO: Self;
> }
> 
> pub struct NonZero<T: Zeroable>(T);
> 
> impl<T: Zeroable> NonZero<T> {
>   const fn new(v: T) -> Option<Self> {
>     if (v == T::ZERO) {
>       None
>     } else {
>       Some(NonZero(v))
>     }
>   }
>   // ...
> }
> ```


Could we instead have it just be an unsafe marker auto trait (without the associated const), and later on have a `WithNiche` trait that'd be a generalization in which the niche may or may not be a zeroed bit pattern? Your proposal seems a bit weird to me, since the compiler *should* be able to create a zeroed bit pattern for any known type. Furthermore, this would allow us to move this particular issue forward while we deal with the complexity of user-defined niches in a separate RFC.

Bikeshed:

```rust
use core::mem::MaybeUninit;

unsafe auto trait ValidZeroed: WithNiche {}
unsafe trait WithNiche {
    const NICHE: Self;
}

unsafe impl<T: ValidZeroed> WithNiche for T {
    const NICHE: T = unsafe { MaybeUninit::zeroed().assume_init() };
}
```

**Comment by reitermarkus** [comment]
> the compiler should be able to create a zeroed bit pattern for any known type

The problem is not *creating* the bit pattern, it's ensuring that pattern is used as a niche, which currently can only be done with `#[rustc_layout_scalar_valid_range_{start,end}]`.

> Furthermore, this would allow us to move this particular issue forward

Creating a user-implementable `Zeroable` trait is not part of this tracking issue, so this is not a blocker in any case.

**Comment by RalfJung** [comment]
Indeed, any discussion of having user-defined NonZero types is off-topic in this tracking issue.

**Comment by tmccombs** [comment]
> Indeed, any discussion of having user-defined NonZero types is off-topic in this tracking issue.

I probably should have been more clear, but the reason I brought it up is I'm not sure if it is merged as is, if it would be possible to change it to allow user-defined NonZero types in the future in a backwards compatible way.
